lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+	dbm "github.com/tendermint/tendermint/libs/db"
+	log "github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/lite"
+	lclient "github.com/tendermint/tendermint/lite/client"
+	lerr "github.com/tendermint/tendermint/lite/errors"
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	loggerPath = "lite"
+	memDBFile  = "trusted.mem"
+	lvlDBFile  = "trusted.lvl"
+	dbName     = "trust-base"
+)
+
+type TrustOptions struct {
+	// Required: only trust commits up to this old.
+	TrustPeriod time.Duration `json:"trust-period"`
+
+	// Option 1: TrustHeight and TrustHash can both be provided
+	// to force the trusting of a particular height and hash.
+	// If the latest trusted height/hash is more recent, then this option is
+	// ignored.
+	TrustHeight int64  `json:"trust-height"`
+	TrustHash   []byte `json:"trust-hash"`
+
+	// Option 2: Callback can be set to implement a confirmation
+	// step if the trust store is uninitialized, or expired.
+	Callback func(height int64, hash []byte) error
+}
+
+// initProvider sets up the databases and loggers and instantiates the Provider object
+func initProvider(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger, cacheSize int, options TrustOptions) *Provider {
+	logger = logger.With("module", loggerPath)
+	logger.Info(fmt.Sprintf("lite/verifying/NewProvider: chainID -> %s, rootDir -> %s, client -> %s", chainID, rootDir, client))
+
+	trust := lite.NewMultiProvider(
+		lite.NewDBProvider(memDBFile, dbm.NewMemDB()).SetLimit(cacheSize),
+		lite.NewDBProvider(lvlDBFile, dbm.NewDB(dbName, dbm.LevelDBBackend, rootDir)),
+	)
+
+	return makeProvider(chainID, options.TrustPeriod, trust, lclient.NewProvider(chainID, client), logger)
+}
+
+// NOTE If you retain the resulting verifier in memory for a long time,
+// usage of the verifier may eventually error, but immediate usage should
+// not error like that, so that e.g. cli usage never errors unexpectedly.
+func NewProvider(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger, cacheSize int, options TrustOptions) (*Provider, error) {
+
+	vp := initProvider(chainID, rootDir, client, logger, cacheSize, options)
+
+	// Get the latest source commit, or the one provided in options.
+	trustCommit, err := getTrustCommit(client, options)
+	if err != nil {
+		return nil, err
+	}
+
+	err = vp.fillValsetAndSaveFC(trustCommit, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// sanity check
+	if time.Now().Sub(trustCommit.Time) <= 0 {
+		panic(fmt.Sprintf("impossible time %v vs %v", time.Now(), trustCommit.Time))
+	}
+
+	// Otherwise we're syncing within the unbonding period.
+	// NOTE: There is a duplication of fetching this latest commit (since
+	// UpdateToHeight() will fetch it again, and latestCommit isn't used), but
+	// it's only once upon initialization of a validator so it's not a big
+	// deal.
+	if options.TrustHeight > 0 {
+		latestCommit, err := client.Commit(nil)
```

Comment:
IMHO, it would be cool if the code mentioned what `client.Commit(nil)` means (the localclient will use the latest stored height if it was given a `nil` height).